src/System.Memory/src/System/Buffers/BuffersExtensions.cs

```diff
@@ -72,14 +72,12 @@ public static void Write<T>(this IBufferWriter<T> bufferWriter, ReadOnlySpan<T>
 
             while (source.Length > 0)
             {
-                int writeSize = destination.Length;
-
                 if (destination.Length == 0)
                 {
-                    writeSize = Math.Min(source.Length, bufferWriter.MaxBufferSize);
-                    destination = bufferWriter.GetSpan(writeSize);
+                    destination = bufferWriter.GetSpan(source.Length);
```

Comment:
From https://github.com/dotnet/corefx/pull/27007#discussion_r168019196

> Aside from the first time through the loop, this will always be true (since `destination = default`). Consider special casing the first iteration outside the loop.

Any thoughts on this?

Reply:
I tried it, looks pretty bad.

Reply:
Seperate the while loop into a different function so this just contains the fast path, then tail calls into the multi write while in else?

Reply:
@benaadams it not exactly about that. Comment is about destination.Length check in every iteration

Reply:
> I tried it, looks pretty bad.

How so? Something like this seems reasonable to me:

```C#
        if (source.Length > 0)
        {
            if (destination.Length == 0)
            {
                destination = bufferWriter.GetSpan(source.Length);
            }

            int writeSize = Math.Min(destination.Length, source.Length);
            source.Slice(0, writeSize).CopyTo(destination);
            bufferWriter.Advance(writeSize);
            source = source.Slice(writeSize);

            while (source.Length > 0)
            {
                destination = bufferWriter.GetSpan(source.Length);

                writeSize = Math.Min(destination.Length, source.Length);
                source.Slice(0, writeSize).CopyTo(destination);
                bufferWriter.Advance(writeSize);
                source = source.Slice(writeSize);
            }
        }
```

Reply:
```csharp
public static void Write<T>(this IBufferWriter<T> bufferWriter, ReadOnlySpan<T> source)
{
    Span<T> destination = bufferWriter.GetSpan();

    // Fast path, try copying to the available memory directly
    if (source.Length <= destination.Length)
    {
        source.CopyTo(destination);
        bufferWriter.Advance(source.Length);
    }
    else
    {
        bufferWriter.WriteMulti(source, destination);
    }
}

private static void WriteMulti<T>(this IBufferWriter<T> bufferWriter, ReadOnlySpan<T> source, Span<T> destination)
{
    if (destination.Length == 0)
    {
        destination = bufferWriter.GetSpan(source.Length);
    }
    
    do
    {
        int writeSize = Math.Min(destination.Length, source.Length);
        source.Slice(0, writeSize).CopyTo(destination);
        bufferWriter.Advance(writeSize);
        source = source.Slice(writeSize);
        
        if (source.Length > 0)
        {
            destination = bufferWriter.GetSpan(source.Length);
            continue;
        }
    }
    while (true);
}
```

Reply:
In your example, @benaadams, WriteMulti has an infinite loop. When do you ever break from it?

Reply:
Changed end to `while (false);` 😄

Reply:
> Also the continue seems redundant.

Don't want to waste time either clearing span as original and rechecking at loop start; or Getting a new one if there is one thing left to write; so it just brings the loop condition internal plus a couple extras.

i.e. if there is more in source (so would loop again) its implicit destination is full so you need a new span and no extra checking is needed

Reply:
> Changed end to `while (false);`

With while(false), I understand why the continue is necessary.

> No need to either clear span (`= default` ); or get a new span if there is nothing thing left to write; so it just brings the loop condition internal plus a getting new span.

I don't understand what you mean. I don't see any unnecessary work in the last iteration of the loop. To me, the continue is necessary for correctness.

Reply:
```csharp
while (source.Length > 0)
{
    if (destination.Length == 0) // unnecessary check other than first iter
                                 // hence move out, and switch to do loop, doing prechecks with if block
    //...
    destination = default;       // unnecessary clear with loop change
}
```

Reply:
Was comparing to current